File: benchmarks/comparison/enum_dispatch.py
def eval_op(tag, a, b):
    if tag == 0:
        return a + b
    elif tag == 1:
        return a * b
    else:
        return -a

total = 0
for i in range(5000000):
    total += eval_op(0, i, i + 1)
    total += eval_op(1, i, 2)
    total += eval_op(2, i, 0)
print(total)

File: benchmarks/comparison/enum_dispatch.c
#include <stdio.h>
#include <stdint.h>

enum OpTag { OP_ADD, OP_MUL, OP_NEG };

typedef struct {
    enum OpTag tag;
    int64_t a, b;
} Op;

static int64_t eval_op(Op op) {
    switch (op.tag) {
        case OP_ADD: return op.a + op.b;
        case OP_MUL: return op.a * op.b;
        case OP_NEG: return -op.a;
    }
    return 0;
}

int main(void) {
    int64_t total = 0;
    for (int64_t i = 0; i < 2000000000; i++) {
        total += eval_op((Op){OP_ADD, i, i + 1});
        total += eval_op((Op){OP_MUL, i, 2});
        total += eval_op((Op){OP_NEG, i, 0});
        total ^= i;
    }
    printf("%ld\n", total);
    return 0;
}
